src/main/java/com/lambdaworks/redis/resource/DefaultClientResources.java

```diff
@@ -316,13 +317,13 @@ public Builder dnsResolver(DnsResolver dnsResolver) {
         }
 
         /**
-         * Sets the reconnect {@link Delay} to delay reconnect attempts. Defaults to binary exponential delay capped at
+         * Sets the reconnect {@link Supplier<Delay>} to delay reconnect attempts. Defaults to binary exponential delay capped at
          * {@literal 30 SECONDS}.
          *
          * @param reconnectDelay the reconnect delay, must not be {@literal null}.
          * @return this
          */
-        public Builder reconnectDelay(Delay reconnectDelay) {
+        public Builder reconnectDelay(Supplier<Delay> reconnectDelay) {
```

Comment:
We should keep the existing method signature and add the `reconnectDelay(Supplier<Delay>)` method. The existing method would call `reconnectDelay(Supplier<Delay>)`. By having an additional/nested interface in `Delay` (declaring the `reset()` method) we could add a check to `reconnectDelay(Delay reconnectDelay)` that checks whether the used delay is stateless or stateful. Stateful delay instances could be rejected to prevent shared usage.